### ISSUE TITLE
[XDS] add XDS health check for Envoy to health check on Config Server

### ISSUE THREAD

**Issue by stevenzzzz** [envoyproxy/envoy#22838]
*Title*: add XDS health check for Envoy to health check on Config Server

*Description*:
With Google GRPC, we are seeing grpc streams hang while the Config server is gone due to machine error. The grpc keepalive mechanism has its own issue that fails to send keepalive pings to config server and make possible remedy(say reconnect). 

To stop such/similar under layer errors, we can build an application layer health check mechanism that does active health check on config server, and reconnect (to possibly a good server) if the being connected config server is malfunctioning/grpc-lib isn't working as expected. 

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

**Comment by stevenzzzz** [comment]
@htuch @adisuissa @yanavlasov

**Comment by htuch** [comment]
Some questions:
* Is this going to be a generic xDS transport health check that works for both Google and Envoy gRPC?
* If this is Google gRPC only, should we aim to do this via gRPC library features? @markdroth

**Comment by adisuissa** [comment]
I think the Envoy gRPC can use the referenced cluster's health-check capabilities (for xDS and non-xDS queries).
One possibility is to add this feature to the Google gRPC configuration, but if it is handled by the underlying gRPC library it would be more generic.

**Comment by stevenzzzz** [comment]
there is a health check in gRPC https://github.com/grpc/grpc/blob/master/doc/keepalive.md

But it's hard to follow the settings there and we are seeing problems with keepalives sent in grpc-lib, that the keepalives could be sent hours apart while the server is long gone. 

We are thinking that if we build an application layer HC mechanism, that gives Envoy users more control over the Envoy-Config-plane-server link, from the client side. 

This would be an feature that's above GRPC, "XDS-local".

**Comment by markdroth** [comment]
If the core problem here is that you're having problems with keepalives, then I think we should work on fixing that rather than inventing a separate mechanism.  I don't think it makes sense to invent yet another application-level health checking mechanism.

If you want to check that gRPC itself is working right, there is already a [health-check protocol](https://github.com/grpc/grpc/blob/master/doc/health-checking.md) for that, and we also have built-in support for [doing this health check from the gRPC client](https://github.com/grpc/proposal/blob/master/A17-client-side-health-checking.md).  But I suspect that does not actually solve the problem you're having here.

If you want to check that the xDS layer is working right, I'm not sure how you do that without actually speaking the xDS transport protocol, which doesn't seem to provide much benefit, since that seems to be exactly what you're having problems with.
